harness/determined/_checkpointing.py

```diff
@@ -0,0 +1,62 @@
+import logging
+from typing import Any, Dict, Optional
+
+import determined as det
+from determined.experimental import client
+
+log = logging.getLogger("generic")
+
+
+class Checkpointing:
+    """
+    Some checkpoint-related REST API wrappers.
+    """
+
+    def __init__(
+        self,
+        session: client.Session,
+        api_path: str,
+        static_metadata: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        self._session = session
+        self._static_metadata = static_metadata or {}
+        self._static_metadata["determined_version"] = det.__version__
+        self._api_path = api_path
+
+    def _report_checkpoint(
+        self,
+        uuid: str,
+        resources: Optional[Dict[str, int]] = None,
+        metadata: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        """
+        After having uploaded a checkpoint, report its existence to the master.
+
+        This is still a private function because it might never be offered directly to the user;
+        they might always go through a StorageManager.
+        """
+        resources = resources or {}
+        metadata = metadata or {}
+        required = {"framework", "format", "total_batches"}
+        allowed = required.union({"total_records", "total_epochs"})
+        missing = [k for k in required if k not in metadata]
```

Comment:
Right, when I wrote this API I was designing it for a future state.  It happens that _right now_ those metadata are required by the system, but they shouldn't be.  And I expect that before we publish the Generic API that we fix that.

For example:
- `trial_id` (part of `static_metadata`): these are required right now, but who says that all checkpoints should be from a trial?  Any Task (in the new meaning of the word) should be checkpointable.
- `trial_run_id` (part of `static_metadata`): Ditto, this should not be required at all.
- `framework`: I'd like for us to stop putting this data into the checkpoint and letting the user do something like `det.pytorch.trial_from_checkpoint_files(...)`.  I don't think it's asking very much for users to know what framework they are loading.  This API is only needed for the `checkpoint.load()` part of checkpoint export, which is not extensible enough to support the Generic API (that is, it will be rethought soon).
- `total_batches`: I think the webui currently requires this to know how to render checkpoints.  But in the future I think this should not be a special value, but simply a value that our python trial workload sequencer saves as metadata.  Or since all of our training APIs are based around `total_batches` maybe it makes sense to let it still be a specially-treated metadata field, but one that is optional.  In any case, this is certainly irrelevant to non-training Tasks in the future.